ext/odbc/php_odbc.c

```diff
@@ -49,23 +51,230 @@
+static inline odbc_result *odbc_result_from_obj(zend_object *obj);
+
+static int le_pconn;
+
+static zend_class_entry *odbc_connection_ce, *odbc_result_ce;
+static zend_object_handlers odbc_connection_object_handlers, odbc_result_object_handlers;
+
+#define Z_ODBC_LINK_P(zv) odbc_link_from_obj(Z_OBJ_P(zv))
+#define Z_ODBC_CONNECTION_P(zv) Z_ODBC_LINK_P(zv)->connection
+#define Z_ODBC_RESULT_P(zv) odbc_result_from_obj(Z_OBJ_P(zv))
+
+static void odbc_insert_new_result(odbc_connection *connection, zval *result)
+{
+	ZEND_ASSERT(Z_TYPE_P(result) == IS_OBJECT);
+#if ZEND_DEBUG
+	ZEND_ASSERT(instanceof_function(Z_OBJCE_P(result), odbc_result_ce));
+#endif
+
+	odbc_result *res = Z_ODBC_RESULT_P(result);
+
+	res->index = connection->results.nNextFreeElement;
+	zend_hash_index_add_new(&connection->results, res->index, result);
+	Z_ADDREF_P(result);
+}
+
+static inline odbc_link *odbc_link_from_obj(zend_object *obj)
+{
+	return (odbc_link *)((char *)(obj) - XtOffsetOf(odbc_link, std));
+}
+
+static int _close_pconn_with_res(zval *zv, void *p)
+{
+	zend_resource *le = Z_RES_P(zv);
+
+	if (le->ptr == p) {
+		return ZEND_HASH_APPLY_REMOVE;
+	}
+
+	return ZEND_HASH_APPLY_KEEP;
+}
+
+static int _close_pconn(zval *zv)
+{
+	zend_resource *le = Z_RES_P(zv);
+	if (le->type == le_pconn) {
+		return ZEND_HASH_APPLY_REMOVE;
+	} else {
+		return ZEND_HASH_APPLY_KEEP;
+	}
+}
+
+/* disconnect, and if it fails, then issue a rollback for any pending transaction (lurcher) */
+static void safe_odbc_disconnect( void *handle )
+{
+	int ret;
+
+	ret = SQLDisconnect( handle );
+	if ( ret == SQL_ERROR )
+	{
+		SQLTransact( NULL, handle, SQL_ROLLBACK );
+		SQLDisconnect( handle );
+	}
+}
+
+static void free_connection(odbc_connection *conn, bool persistent)
+{
+	/* If aborted via timer expiration, don't try to call any unixODBC function */
+	if (!(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
+		safe_odbc_disconnect(conn->hdbc);
+		SQLFreeConnect(conn->hdbc);
+		SQLFreeEnv(conn->henv);
+	}
+
+	conn->hdbc = NULL;
+	conn->henv = NULL;
+
+	zend_hash_destroy(&conn->results);
+
+	pefree(conn, persistent);
+
+	ODBCG(num_links)--;
+	if (persistent) {
+		ODBCG(num_persistent)--;
+	}
+}
+
+static void odbc_link_free(odbc_link *link)
+{
+	ZEND_ASSERT(link->connection && "link has already been closed");
+
+	close_results_with_connection(link->connection);
+
+	if (!link->persistent) {
+		free_connection(link->connection, link->persistent);
+	}
+
+	link->connection = NULL;
+
+	if (link->hash) {
+		zend_hash_del(&ODBCG(connections), link->hash);
+		zend_string_release_ex(link->hash, link->persistent);
+		link->hash = NULL;
+	}
+}
+
+static zend_object *odbc_connection_create_object(zend_class_entry *class_type)
+{
+	odbc_link *intern = zend_object_alloc(sizeof(odbc_link), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &odbc_connection_object_handlers;
```

Comment:
This line isn't necessary I think because the default object handler is set in the ce.